# digits

In [1]:
from sklearn.datasets import load_digits
digits = load_digits()
import pandas as pd

df = pd.DataFrame(data=digits.data, columns=digits.feature_names)
df.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [2]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1797 entries, 0 to 1796
Data columns (total 64 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pixel_0_0  1797 non-null   float64
 1   pixel_0_1  1797 non-null   float64
 2   pixel_0_2  1797 non-null   float64
 3   pixel_0_3  1797 non-null   float64
 4   pixel_0_4  1797 non-null   float64
 5   pixel_0_5  1797 non-null   float64
 6   pixel_0_6  1797 non-null   float64
 7   pixel_0_7  1797 non-null   float64
 8   pixel_1_0  1797 non-null   float64
 9   pixel_1_1  1797 non-null   float64
 10  pixel_1_2  1797 non-null   float64
 11  pixel_1_3  1797 non-null   float64
 12  pixel_1_4  1797 non-null   float64
 13  pixel_1_5  1797 non-null   float64
 14  pixel_1_6  1797 non-null   float64
 15  pixel_1_7  1797 non-null   float64
 16  pixel_2_0  1797 non-null   float64
 17  pixel_2_1  1797 non-null   float64
 18  pixel_2_2  1797 non-null   float64
 19  pixel_2_3  1797 non-null   float64
 20  pixel_2_

In [4]:
df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_6    pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      3.725097     0.206455     0.000556     0.279354     5.557596   
std       4.919406     0.984401     0.023590     0.934302     5.103019   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     1.000000   
50%       1.000000     0.000000     0.000000     0.000000     4.000000   
75%       7.000000     0.000000     0.000000     0.000000    10.000000   
max      16.000000    13.000000     1.000000     9.000000    16.000000   

         pixel_7_3    pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     12.089037    11.809126     6.764051     2.067891     0.364496  
std       4.374694     4.933947     5.900623     4.090548     1.860122  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      11.000000    10.000000     0.000000     0.000000     0.000000  
50%      13.000000    14.000000     6.000000     0.000000     0.000000  
75%      16.000000    16.000000    12.000000     2.000000     0.000000  
max      16.000000    16.000000    16.000000    16.000000    16.000000  

[8 rows x 64 columns]

In [5]:
# 정규화
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
digits_scaled = scaler.fit_transform(digits.data)
scaled_df = pd.DataFrame(data = digits_scaled, columns = digits.feature_names)
scaled_df.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0     0.3125     0.8125     0.5625     0.0625   
1        0.0        0.0     0.0000     0.7500     0.8125     0.3125   
2        0.0        0.0     0.0000     0.2500     0.9375     0.7500   
3        0.0        0.0     0.4375     0.9375     0.8125     0.0625   
4        0.0        0.0     0.0000     0.0625     0.6875     0.0000   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...     0.0000        0.0   
1        0.0        0.0        0.0        0.0  ...     0.0000        0.0   
2        0.0        0.0        0.0        0.0  ...     0.3125        0.0   
3        0.0        0.0        0.0        0.5  ...     0.5625        0.0   
4        0.0        0.0        0.0        0.0  ...     0.0000        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0     0.3750     0.8125     0.6250     0.0000   
1        0.0        0.0     0.0000     0.6875     1.0000     0.6250   
2        0.0        0.0     0.0000     0.1875     0.6875     1.0000   
3        0.0        0.0     0.4375     0.8125     0.8125     0.5625   
4        0.0        0.0     0.0000     0.1250     1.0000     0.2500   

   pixel_7_6  pixel_7_7  
0     0.0000        0.0  
1     0.0000        0.0  
2     0.5625        0.0  
3     0.0000        0.0  
4     0.0000        0.0  

[5 rows x 64 columns]

In [6]:
# 트레인셋, 테스트셋
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    digits.data, digits.target, test_size = 0.2, random_state = 2021
)

In [7]:
# 하이퍼파라미터

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

dt = DecisionTreeClassifier()
rf = RandomForestClassifier()
svc = SVC()
lr = LogisticRegression()
knn = KNeighborsClassifier()

In [8]:
dt.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': 'deprecated',
 'random_state': None,
 'splitter': 'best'}

In [9]:
params = {
    'min_samples_split': [2, 3, 4],
    'max_depth': [14, 16, 18]
}

In [10]:
from sklearn.model_selection import GridSearchCV
grid_dt = GridSearchCV(dt, param_grid = params, scoring='accuracy', cv=5)
grid_dt.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [14, 16, 18],
                         'min_samples_split': [2, 3, 4]},
             scoring='accuracy')

In [11]:
print('GridSearchCV 최적 파라미터:', grid_dt.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_dt.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 14, 'min_samples_split': 3}
GridSearchCV 최고 정확도: 0.8560


In [12]:
best_dt = grid_dt.best_estimator_
best_dt

DecisionTreeClassifier(max_depth=14, min_samples_split=3)

In [13]:
rf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [14]:
params = {
    'min_samples_split': [3, 4, 5],
    'max_depth': [12, 16, 20]
}

In [15]:
from sklearn.model_selection import GridSearchCV
grid_rf = GridSearchCV(rf, param_grid = params, scoring='accuracy', cv=5)
grid_rf.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'max_depth': [12, 16, 20],
                         'min_samples_split': [3, 4, 5]},
             scoring='accuracy')

In [16]:
print('GridSearchCV 최적 파라미터:', grid_rf.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_rf.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 12, 'min_samples_split': 4}
GridSearchCV 최고 정확도: 0.9756


In [17]:
best_rf = grid_rf.best_estimator_
best_rf

RandomForestClassifier(max_depth=12, min_samples_split=4)

In [18]:
# svc 하이퍼파라미터
svc.get_params()

{'C': 1.0,
 'break_ties': False,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 'scale',
 'kernel': 'rbf',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [19]:
params = {
    'C': [0.9, 1.0, 1.1, 1.4]
}

In [20]:
from sklearn.model_selection import GridSearchCV
grid_svc = GridSearchCV(svc, param_grid = params, scoring='accuracy', cv=5)
grid_svc.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=SVC(), param_grid={'C': [0.9, 1.0, 1.1, 1.4]},
             scoring='accuracy')

In [21]:
print('GridSearchCV 최적 파라미터:', grid_svc.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_svc.best_score_))

GridSearchCV 최적 파라미터: {'C': 1.1}
GridSearchCV 최고 정확도: 0.9896


In [22]:
best_svc = grid_rf.best_estimator_
best_svc

RandomForestClassifier(max_depth=12, min_samples_split=4)

In [23]:
# lr하이퍼 파라미터
lr.get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'auto',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [24]:
params = {
    'C': [0.9, 1.0, 1.1, 1.2, 1.3, 1.4]
}

In [25]:
from sklearn.model_selection import GridSearchCV
grid_lr = GridSearchCV(lr, param_grid = params, scoring='accuracy', cv=5)
grid_lr.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.9, 1.0, 1.1, 1.2, 1.3, 1.4]},
             scoring='accuracy')

In [26]:
print('GridSearchCV 최적 파라미터:', grid_lr.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_lr.best_score_))

GridSearchCV 최적 파라미터: {'C': 0.9}
GridSearchCV 최고 정확도: 0.9638


In [27]:
best_lr = grid_lr.best_estimator_
best_lr

LogisticRegression(C=0.9)

In [28]:
# knn 하이퍼파라미터
knn.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [29]:
params = {
    'leaf_size': [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]
}

In [30]:
from sklearn.model_selection import GridSearchCV
grid_knn = GridSearchCV(knn, param_grid = params, scoring='accuracy', cv=5)
grid_knn.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'leaf_size': [20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
                                       30]},
             scoring='accuracy')

In [31]:
print('GridSearchCV 최적 파라미터:', grid_knn.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_knn.best_score_))

GridSearchCV 최적 파라미터: {'leaf_size': 20}
GridSearchCV 최고 정확도: 0.9819


In [32]:
best_knn = grid_knn.best_estimator_
best_knn

KNeighborsClassifier(leaf_size=20)

In [34]:
# 앙상블
from sklearn.ensemble import VotingClassifier

vo_clf = VotingClassifier(
    estimators = [('DT',best_dt), ('RF',best_rf), ('SVC',best_svc), ('KNN',best_knn)],
    voting = 'hard'
)

vo_clf.fit(X_train, y_train)
pred = vo_clf.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.9638888888888889

# Breast_Cancer

In [35]:
from sklearn.datasets import load_breast_cancer
cancer = load_breast_cancer()

In [36]:
import pandas as pd

df = pd.DataFrame(data=cancer.data, columns=cancer.feature_names)
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [37]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [38]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    cancer.data,cancer.target, test_size=0.2, random_state=2021
)

In [39]:
# 정규화
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
digits_scaled = scaler.fit_transform(digits.data)
scaled_df = pd.DataFrame(data = digits_scaled, columns = digits.feature_names)
scaled_df.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0     0.3125     0.8125     0.5625     0.0625   
1        0.0        0.0     0.0000     0.7500     0.8125     0.3125   
2        0.0        0.0     0.0000     0.2500     0.9375     0.7500   
3        0.0        0.0     0.4375     0.9375     0.8125     0.0625   
4        0.0        0.0     0.0000     0.0625     0.6875     0.0000   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...     0.0000        0.0   
1        0.0        0.0        0.0        0.0  ...     0.0000        0.0   
2        0.0        0.0        0.0        0.0  ...     0.3125        0.0   
3        0.0        0.0        0.0        0.5  ...     0.5625        0.0   
4        0.0        0.0        0.0        0.0  ...     0.0000        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0     0.3750     0.8125     0.6250     0.0000   
1        0.0        0.0     0.0000     0.6875     1.0000     0.6250   
2        0.0        0.0     0.0000     0.1875     0.6875     1.0000   
3        0.0        0.0     0.4375     0.8125     0.8125     0.5625   
4        0.0        0.0     0.0000     0.1250     1.0000     0.2500   

   pixel_7_6  pixel_7_7  
0     0.0000        0.0  
1     0.0000        0.0  
2     0.5625        0.0  
3     0.0000        0.0  
4     0.0000        0.0  

[5 rows x 64 columns]

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

dt = DecisionTreeClassifier()
rf = RandomForestClassifier()
lr = LogisticRegression()
svc = SVC()
knn = KNeighborsClassifier()

In [40]:
# dt 하이퍼파라미터
params = {
    'min_samples_split': [4, 5, 6],
    'max_depth': [10, 12, 14]
}

In [41]:
from sklearn.model_selection import GridSearchCV
grid_dt = GridSearchCV(dt, param_grid = params, scoring='accuracy', cv=5)
print(grid_dt.fit(X_train, y_train))

print('GridSearchCV 최적 파라미터:', grid_dt.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_dt.best_score_))

best_dt = grid_dt.best_estimator_
print(best_dt)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [10, 12, 14],
                         'min_samples_split': [4, 5, 6]},
             scoring='accuracy')
GridSearchCV 최적 파라미터: {'max_depth': 10, 'min_samples_split': 4}
GridSearchCV 최고 정확도: 0.9275
DecisionTreeClassifier(max_depth=10, min_samples_split=4)


In [42]:
# rf 하이퍼파라미터
params = {
    'min_samples_split': [3, 4, 5],
    'max_depth': [16, 20, 24]
}

In [43]:
from sklearn.model_selection import GridSearchCV
grid_rf = GridSearchCV(rf, param_grid = params, scoring='accuracy', cv=5)
print(grid_rf.fit(X_train, y_train))

print('GridSearchCV 최적 파라미터:', grid_rf.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_rf.best_score_))

best_rf = grid_rf.best_estimator_
print(best_rf)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'max_depth': [16, 20, 24],
                         'min_samples_split': [3, 4, 5]},
             scoring='accuracy')
GridSearchCV 최적 파라미터: {'max_depth': 16, 'min_samples_split': 4}
GridSearchCV 최고 정확도: 0.9582
RandomForestClassifier(max_depth=16, min_samples_split=4)


In [44]:
# lr 하이퍼파라미터
params = {
    'C': [0.8, 0.9, 1.0, 1.1, 1.2]
}

In [45]:
from sklearn.model_selection import GridSearchCV
grid_lr = GridSearchCV(lr, param_grid = params, scoring='accuracy', cv=5)
print(grid_lr.fit(X_train, y_train))

print('GridSearchCV 최적 파라미터:', grid_lr.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_lr.best_score_))

best_lr = grid_lr.best_estimator_
print(best_lr)

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.8, 0.9, 1.0, 1.1, 1.2]}, scoring='accuracy')
GridSearchCV 최적 파라미터: {'C': 1.1}
GridSearchCV 최고 정확도: 0.9495
LogisticRegression(C=1.1)


In [46]:
# svc 하이퍼파라미터
params = {
    'C': [0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2]
}

In [47]:
from sklearn.model_selection import GridSearchCV
grid_svc = GridSearchCV(svc, param_grid = params, scoring='accuracy', cv=5)
print(grid_svc.fit(X_train, y_train))

print('GridSearchCV 최적 파라미터:', grid_svc.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_svc.best_score_))

best_svc = grid_svc.best_estimator_
print(best_svc)

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2]},
             scoring='accuracy')
GridSearchCV 최적 파라미터: {'C': 1.2}
GridSearchCV 최고 정확도: 0.9253
SVC(C=1.2)


In [48]:
# knn 하이퍼파라미터
params = {
    'leaf_size': [20, 22, 24, 26, 28, 30]
}

In [49]:
from sklearn.model_selection import GridSearchCV
grid_knn = GridSearchCV(knn, param_grid = params, scoring='accuracy', cv=5)
print(grid_knn.fit(X_train, y_train))

print('GridSearchCV 최적 파라미터:', grid_knn.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_knn.best_score_))

best_knn = grid_knn.best_estimator_
print(best_knn)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'leaf_size': [20, 22, 24, 26, 28, 30]},
             scoring='accuracy')
GridSearchCV 최적 파라미터: {'leaf_size': 20}
GridSearchCV 최고 정확도: 0.9275
KNeighborsClassifier(leaf_size=20)


In [50]:
# 앙상블
from sklearn.ensemble import VotingClassifier

vo_clf = VotingClassifier(
    estimators = [('DT',best_dt), ('RF',best_rf), ('LR',lr), ('SVC',svc), ('KNN',knn)],
    voting = 'hard'
)

vo_clf.fit(X_train, y_train)
pred = vo_clf.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.9210526315789473

digits
정확도는 RandomForestClassifier(max_depth=12, min_samples_split=4)일때 0.9896로 가장 높음

유방암
정확도는 RandomForestClassifier(max_depth=16, min_samples_split=4)일때 0.9582로 가장 높음
